packages/core/src/v3-compat.ts

```diff
@@ -0,0 +1,318 @@
+    }
+}
+
+/**
+ * Parameters for creating a new [[Session]] through [[API.createSession]]
+ */
+export class CreateSessionParams extends Serializable {
+    /** The id of the [[Account]] to create the session for */
+    account: AccountID = "";
+
+    /** Verification value used for SRP session negotiation */
+    @AsBytes()
+    M!: Uint8Array;
+
+    /** Random value used form SRP session negotiation */
+    @AsBytes()
+    A!: Uint8Array;
+
+    constructor(props?: Partial<CreateSessionParams>) {
+        super();
+        props && Object.assign(this, props);
+    }
+}
+
+const srpIds = new Map<AccountID, string>();
+
+type Constructor<T> = new (...args: any[]) => T;
+
+/**
+ * V3 Compatibility Layer, implemented as a mixin
+ * @deprecated
+ */
+export const V3Compat = (base: Constructor<Controller>) => {
+    class C extends base {
+        /**
+         * Request verification of a given email address. This will send a verification code
+         * to the email in question which can then be exchanged for a verification token via
+         * [[completeEmailVerification]].
+         * @deprecated since v4.0. Please use [[startAuthRequest]] instead
+         */
+        @Handler(RequestMFACodeParams, undefined)
+        async requestMFACode({ email, purpose }: RequestMFACodeParams) {
+            await this.startAuthRequest(
+                new StartAuthRequestParams({ email, purpose: mapV3AuthPurpose(purpose), type: AuthType.Email })
+            );
+        }
+
+        /**
+         * Complete the email verification process by providing a verification code received
+         * via email. Returns a verification token that can be used in other api calls like
+         * [[createAccount]] or [[recoverAccount]].
+         * @deprecated since v4.0. Please use [[completeAuthRequest]] instead
+         */
+        @Handler(RetrieveMFATokenParams, RetrieveMFATokenResponse)
+        async retrieveMFAToken({ email, code, purpose: legacyPurpose }: RetrieveMFATokenParams) {
+            const purpose = mapV3AuthPurpose(legacyPurpose);
+            try {
+                const auth = await this._getAuth(email);
+
+                const request = auth.authRequests.find(
+                    (m) =>
+                        m.type === AuthType.Email &&
+                        m.purpose === purpose &&
+                        m.state.email === email &&
+                        m.state.verificationCode === code
+                );
+                if (!request) {
+                    throw new Err(ErrorCode.AUTHENTICATION_FAILED, "Failed to complete auth request.");
+                }
+
+                if (request.tries >= 3) {
+                    throw new Err(
+                        ErrorCode.AUTHENTICATION_TRIES_EXCEEDED,
+                        "You have exceed your allowed numer of tries!"
+                    );
+                }
+
+                const authenticators = await this._getAuthenticators(auth);
```

Comment:
This is the step you were missing when trying to get the `retrieveMFAToken` endpoint to work. Some authenticators don't need to be registered by the users but are added "ad hoc". By default the email authenticator is one of them. The catch is that you have to use `_getAuthenticators` to have these "ad hoc" authenticators included.